File: tests/helpers/memoryDb.ts
/**
 * In-memory Db for integration tests.
 *
 * Implements the exact query strings used by AuthService (src/api/authService.ts)
 * and UsageService (src/api/usage.ts) so endpoint tests exercise the full
 * request → handler → service → SQL path without a real database.
 */
import { randomUUID } from 'node:crypto';
import type { Db } from '../../src/api/db.ts';

export interface MemUser {
  id: string;
  email: string;
  display_name: string | null;
  password_hash: string | null;
  is_beta?: boolean;
}

export interface MemSession {
  user_id: string;
  token_hash: string;
  device: string;
  expires_at: string;
}

export interface MemRun {
  id: number;
  client_run_id: string | null;
  user_id: string;
  model: string;
  input_tokens: number;
  output_tokens: number;
}

export interface MemUsageRow {
  id: number;
  user_id: string;
  run_id: number;
  model: string;
  input_tokens: number;
  output_tokens: number;
  created_at: string;
}

export interface MemFreeSession {
  id: string;
  user_id: string;
  /** UTC day bucket (YYYY-MM-DD) the session belongs to. */
  day: string;
  session_number: number;
  started_at: string;
  expires_at: string;
  ended_at: string | null;
}

export interface MemDailyCost {
  user_id: string;
  day: string;
  spent: number;
  reserved: number;
  version: number;
}

export interface MemGlobalCost {
  period_type: string;
  period: string;
  spent: number;
  reserved: number;
  version: number;
}

export interface MemAiUsage {
  user_id: string;
  session_id: string | null;
  provider: string;
  model: string;
  input_tokens: number;
  cached_input_tokens: number;
  output_tokens: number;
  total_tokens: number;
  estimated_cost_usd_micros: number;
  day: string;
}

export interface MemoryDb {
  db: Db;
  users: MemUser[];
  sessions: MemSession[];
  runs: MemRun[];
  usageRows: MemUsageRow[];
  freeSessions: MemFreeSession[];
  dailyCosts: MemDailyCost[];
  globalCosts: MemGlobalCost[];
  aiUsage: MemAiUsage[];
}

/** Create a fresh in-memory database. */
export function createMemoryDb(): MemoryDb {
  let nextId = 1;
  const users: MemUser[] = [];
  const sessions: MemSession[] = [];
  const runs: MemRun[] = [];
  const usageRows: MemUsageRow[] = [];
  const freeSessions: MemFreeSession[] = [];
  const dailyCosts: MemDailyCost[] = [];
  const globalCosts: MemGlobalCost[] = [];
  const aiUsage: MemAiUsage[] = [];

  const db: Db = async (sql, params: unknown[] = []) => {
    // Health probe.
    if (sql.includes('SELECT 1')) return [];

    // ---- AuthService ------------------------------------------------------
    if (sql.includes('INSERT INTO users')) {
      const [email, displayName, passwordHash, isBeta] = params as [string, string | null, string, boolean | undefined];
      users.push({
        id: randomUUID(),
        email,
        display_name: displayName,
        password_hash: passwordHash,
        is_beta: isBeta ?? false,
      });
      return [{ id: users[users.length - 1]?.id }];
    }
    if (sql.includes('FROM users') && sql.includes('password_hash')) {
      const email = params[0] as string;
      const u = users.find((x) => x.email === email);
      return u ? [{ id: u.id, email: u.email, display_name: u.display_name, password_hash: u.password_hash }] : [];
    }
    if (sql.includes('SELECT id FROM users')) {
      const email = params[0] as string;
      const u = users.find((x) => x.email === email);
      return u ? [{ id: u.id }] : [];
    }
    if (sql.includes('INSERT INTO sessions')) {
      const [userId, tokenHash, device, expiresAt] = params as [string, string, string, string];
      sessions.push({ user_id: userId, token_hash: tokenHash, device, expires_at: expiresAt });
      return [{ id: nextId++ }];
    }
    if (sql.includes('DELETE FROM sessions')) {
      const tokenHash = params[0] as string;
      const i = sessions.findIndex((s) => s.token_hash === tokenHash);
      if (i >= 0) sessions.splice(i, 1);
      return [];
    }
    if (sql.includes('JOIN users')) {
      const tokenHash = params[0] as string;
      const s = sessions.find((x) => x.token_hash === tokenHash);
      if (!s) return [];
      const u = users.find((x) => x.id === s.user_id);
      if (!u) return [];
      return [{ id: u.id, email: u.email, display_name: u.display_name, expires_at: s.expires_at }];
    }

    // ---- UsageService -----------------------------------------------------
    if (sql.includes('INSERT INTO agent_runs')) {
      const clientRunId = params[0] as string | null;
      if (clientRunId && runs.some((r) => r.client_run_id === clientRunId)) return []; // ON CONFLICT DO NOTHING
      const run: MemRun = {
        id: nextId++,
        client_run_id: clientRunId,
        user_id: params[1] as string,
        model: params[6] as string,
        input_tokens: params[9] as number,
        output_tokens: params[10] as number,
      };
      runs.push(run);
      return [{ id: run.id }];
    }
    if (sql.includes('SELECT id FROM agent_runs')) {
      const clientRunId = params[0] as string;
      const r = runs.find((x) => x.client_run_id === clientRunId);
      return r ? [{ id: r.id }] : [];
    }
    if (sql.includes('INSERT INTO usage')) {
      const [userId, runId, model, inputTokens, outputTokens] = params as [string, number, string, number, number];
      usageRows.push({
        id: nextId++,
        user_id: userId,
        run_id: runId,
        model,
        input_tokens: inputTokens,
        output_tokens: outputTokens,
        created_at: new Date().toISOString(),
      });
      return [];
    }
    if (sql.includes('FROM usage') && sql.includes('WHERE u.user_id')) {
      const userId = params[0] as string;
      const limit = Number(params[1] ?? 20);
      return usageRows
        .filter((r) => r.user_id === userId)
        .slice(0, limit)
        .map((r) => ({
          id: r.id,
          user_id: r.user_id,
          run_id: r.run_id,
          model: r.model,
          input_tokens: r.input_tokens,
          output_tokens: r.output_tokens,
          created_at: r.created_at,
        }));
    }

    // ---- FreeSessionService (GRACE FREE, Milestone 13) --------------------
    if (sql.includes('INSERT INTO free_sessions')) {
      const [userId, day, sessionNumber, startedAt, expiresAt] = params as [
        string, string, number, string, string,
      ];
      // Mirrors the real UNIQUE (user_id, day, session_number) constraint:
      // concurrent starts collide here and the service retries.
      const conflict = freeSessions.some(
        (s) => s.user_id === userId && s.day === day && s.session_number === sessionNumber,
      );
      if (conflict) {
        const err = new Error('duplicate key value violates unique constraint "free_sessions"') as Error & {
          code: string;
        };
        err.code = '23505'; // SQLSTATE unique_violation, like Postgres
        throw err;
      }
      const row: MemFreeSession = {
        id: randomUUID(),
        user_id: userId,
        day,
        session_number: sessionNumber,
        started_at: startedAt,
        expires_at: expiresAt,
        ended_at: null,
      };
      freeSessions.push(row);
      return [{
        id: row.id,
        user_id: row.user_id,
        day: row.day,
        session_number: row.session_number,
        started_at: row.started_at,
        expires_at: row.expires_at,
        ended_at: row.ended_at,
      }];
    }
    if (sql.includes('FROM free_sessions')) {
      const userId = params[0] as string;
      const day = params[1] as string;
      return freeSessions
        .filter((s) => s.user_id === userId && s.day === day)
        .sort((a, b) => a.session_number - b.session_number)
        .map((s) => ({
          id: s.id,
          user_id: s.user_id,
          day: s.day,
          session_number: s.session_number,
          started_at: s.started_at,
          expires_at: s.expires_at,
          ended_at: s.ended_at,
        }));
    }
    if (sql.includes('UPDATE free_sessions')) {
      const id = params[0] as string;
      const row = freeSessions.find((s) => s.id === id);
      if (row && !row.ended_at) {
        // Lazy end (markEnded) passes only the id → ended_at = expires_at.
        // Explicit end (POST /api/session/end) passes the end timestamp as $2.
        row.ended_at = (params[1] as string | undefined) ?? row.expires_at;
      }
      return [];
    }

    // ---- CostGuardService (cost accounting) --------------------------------
    // Reserve: INSERT ... ON CONFLICT (user_id, day) DO UPDATE ... WHERE
    //   spent + reserved + new <= cap. Mirrors the atomic Postgres semantics:
    //   the WHERE clause is re-checked under the (single-threaded) write, so
    //   concurrent requests can never overshoot the ceiling.
    if (sql.includes('INSERT INTO daily_cost')) {
      const [userId, day, micros, capMicros] = params as [string, string, number, number];
      const row = dailyCosts.find((r) => r.user_id === userId && r.day === day);
      if (row) {
        if (row.spent + row.reserved + micros > capMicros) return []; // WHERE false → no row
        row.reserved += micros;
        row.version += 1;
        return [{ user_id: userId }];
      }
      dailyCosts.push({ user_id: userId, day, spent: 0, reserved: micros, version: 1 });
      return [{ user_id: userId }];
    }
    if (sql.includes('UPDATE daily_cost')) {
      // Settle passes 4 params [user, day, spentDelta, reservedDelta]; the
      // release path passes 3 [user, day, reservedDelta].
      const [userId, day] = params as [string, string];
      const isSettle = params.length >= 4;
      const spentDelta = isSettle ? (params[2] as number) : 0;
      const reservedDelta = (isSettle ? (params[3] as number) : (params[2] as number)) ?? 0;
      const row = dailyCosts.find((r) => r.user_id === userId && r.day === day);
      if (row) {
        row.spent += spentDelta;
        row.reserved = Math.max(0, row.reserved - reservedDelta);
        row.version += 1;
      }
      return [];
    }
    if (sql.includes('FROM daily_cost')) {
      const [userId, day] = params as [string, string];
      const row = dailyCosts.find((r) => r.user_id === userId && r.day === day);
      return row ? [{ spent_usd_micros: row.spent, reserved_usd_micros: row.reserved }] : [];
    }
    if (sql.includes('INSERT INTO global_cost')) {
      const [periodType, period, micros, capMicros] = params as [string, string, number, number];
      const row = globalCosts.find((r) => r.period_type === periodType && r.period === period);
      if (row) {
        if (row.spent + row.reserved + micros > capMicros) return [];
        row.reserved += micros;
        row.version += 1;
        return [{ period_type: periodType }];
      }
      globalCosts.push({ period_type: periodType, period, spent: 0, reserved: micros, version: 1 });
      return [{ period_type: periodType }];
    }
    if (sql.includes('UPDATE global_cost')) {
      const [periodType, period] = params as [string, string];
      const isSettle = params.length >= 4;
      const spentDelta = isSettle ? (params[2] as number) : 0;
      const reservedDelta = (isSettle ? (params[3] as number) : (params[2] as number)) ?? 0;
      const row = globalCosts.find((r) => r.period_type === periodType && r.period === period);
      if (row) {
        row.spent += spentDelta;
        row.reserved = Math.max(0, row.reserved - reservedDelta);
        row.version += 1;
      }
      return [];
    }
    if (sql.includes('FROM global_cost')) {
      const [periodType, period] = params as [string, string];
      const row = globalCosts.find((r) => r.period_type === periodType && r.period === period);
      return row ? [{ spent_usd_micros: row.spent, reserved_usd_micros: row.reserved }] : [];
    }
    if (sql.includes('INSERT INTO ai_usage')) {
      const [userId, sessionId, provider, model, input, cached, output, total, cost, day] = params as [
        string, string | null, string, string, number, number, number, number, number, string,
      ];
      aiUsage.push({
        user_id: userId,
        session_id: sessionId,
        provider,
        model,
        input_tokens: input,
        cached_input_tokens: cached,
        output_tokens: output,
        total_tokens: total,
        estimated_cost_usd_micros: cost,
        day,
      });
      return [];
    }

    throw new Error(`MemoryDb: unhandled query: ${sql}`);
  };

  return { db, users, sessions, runs, usageRows, freeSessions, dailyCosts, globalCosts, aiUsage };
}

File: tests/helpers/memory_db.py
"""In-memory Db for backend integration tests (port of tests/helpers/memoryDb.ts).

Implements the exact query strings used by AuthService, UsageService,
FreeSessionService and CostGuardService so endpoint tests exercise the full
request → handler → service → SQL path without a real database. The SQL and
parameter ORDER match the Python backend's services 1:1.
"""

import uuid


class MemUniqueViolation(Exception):
    """Mirrors a Postgres UNIQUE violation (SQLSTATE 23505)."""

    def __init__(self, message: str) -> None:
        super().__init__(message)
        self.code = "23505"
        self.sqlstate = "23505"


class MemUnhandledQuery(Exception):
    pass


def create_memory_db():
    state = {
        "next_id": 1,
        "users": [],
        "sessions": [],
        "runs": [],
        "usage_rows": [],
        "free_sessions": [],
        "daily_costs": [],
        "global_costs": [],
        "ai_usage": [],
    }

    def db(sql: str, params: list | None = None):
        params = params or []

        # Health probe.
        if "SELECT 1" in sql:
            return []

        # ---- AuthService ---------------------------------------------------
        if "INSERT INTO users" in sql:
            email, display_name, password_hash = params[0], params[1], params[2]
            is_beta = params[3] if len(params) > 3 else False
            state["users"].append(
                {
                    "id": str(uuid.uuid4()),
                    "email": email,
                    "display_name": display_name,
                    "password_hash": password_hash,
                    "is_beta": bool(is_beta),
                }
            )
            return [{"id": state["users"][-1]["id"]}]
        if "FROM users" in sql and "password_hash" in sql:
            email = params[0]
            u = next((x for x in state["users"] if x["email"] == email), None)
            return (
                [{"id": u["id"], "email": u["email"], "display_name": u["display_name"], "password_hash": u["password_hash"]}]
                if u
                else []
            )
        if "SELECT id FROM users" in sql:
            email = params[0]
            u = next((x for x in state["users"] if x["email"] == email), None)
            return [{"id": u["id"]}] if u else []
        if "INSERT INTO sessions" in sql:
            user_id, token_hash, device, expires_at = params[0], params[1], params[2], params[3]
            state["sessions"].append(
                {"id": str(uuid.uuid4()), "user_id": user_id, "token_hash": token_hash, "device": device, "expires_at": expires_at}
            )
            return [{"id": state["next_id"]}]
        if "DELETE FROM sessions" in sql:
            token_hash = params[0]
            state["sessions"] = [s for s in state["sessions"] if s["token_hash"] != token_hash]
            return []
        if "JOIN users" in sql:
            token_hash = params[0]
            s = next((x for x in state["sessions"] if x["token_hash"] == token_hash), None)
            if not s:
                return []
            u = next((x for x in state["users"] if x["id"] == s["user_id"]), None)
            if not u:
                return []
            return [{"id": u["id"], "email": u["email"], "display_name": u["display_name"], "expires_at": s["expires_at"]}]

        # ---- UsageService --------------------------------------------------
        if "INSERT INTO agent_runs" in sql:
            client_run_id = params[0]
            if client_run_id and any(r["client_run_id"] == client_run_id for r in state["runs"]):
                return []  # ON CONFLICT DO NOTHING
            run = {
                "id": state["next_id"],
                "client_run_id": client_run_id,
                "user_id": params[1],
                "model": params[6],
                "input_tokens": params[9],
                "output_tokens": params[10],
            }
            state["next_id"] += 1
            state["runs"].append(run)
            return [{"id": run["id"]}]
        if "SELECT id FROM agent_runs" in sql:
            client_run_id = params[0]
            r = next((x for x in state["runs"] if x["client_run_id"] == client_run_id), None)
            return [{"id": r["id"]}] if r else []
        if "INSERT INTO usage" in sql:
            user_id, run_id, model, input_tokens, output_tokens = params[0], params[1], params[2], params[3], params[4]
            state["usage_rows"].append(
                {
                    "id": state["next_id"],
                    "user_id": user_id,
                    "run_id": run_id,
                    "model": model,
                    "input_tokens": input_tokens,
                    "output_tokens": output_tokens,
                    "created_at": "2026-01-01T00:00:00.000Z",
                }
            )
            state["next_id"] += 1
            return []
        if "FROM usage" in sql and "WHERE u.user_id" in sql:
            user_id = params[0]
            limit = int(params[1] or 20)
            return [
                {
                    "id": r["id"],
                    "user_id": r["user_id"],
                    "run_id": r["run_id"],
                    "model": r["model"],
                    "input_tokens": r["input_tokens"],
                    "output_tokens": r["output_tokens"],
                    "created_at": r["created_at"],
                }
                for r in state["usage_rows"]
                if r["user_id"] == user_id
            ][:limit]

        # ---- FreeSessionService -------------------------------------------
        if "INSERT INTO free_sessions" in sql:
            user_id, day, session_number, started_at, expires_at = params[0], params[1], params[2], params[3], params[4]
            conflict = any(
                s["user_id"] == user_id and s["day"] == day and s["session_number"] == session_number
                for s in state["free_sessions"]
            )
            if conflict:
                raise MemUniqueViolation('duplicate key value violates unique constraint "free_sessions"')
            row = {
                "id": str(uuid.uuid4()),
                "user_id": user_id,
                "day": day,
                "session_number": session_number,
                "started_at": started_at,
                "expires_at": expires_at,
                "ended_at": None,
            }
            state["free_sessions"].append(row)
            return [
                {
                    "id": row["id"],
                    "user_id": row["user_id"],
                    "day": row["day"],
                    "session_number": row["session_number"],
                    "started_at": row["started_at"],
                    "expires_at": row["expires_at"],
                    "ended_at": row["ended_at"],
                }
            ]
        if "FROM free_sessions" in sql:
            user_id, day = params[0], params[1]
            rows = [s for s in state["free_sessions"] if s["user_id"] == user_id and s["day"] == day]
            rows.sort(key=lambda s: s["session_number"])
            return [
                {
                    "id": s["id"],
                    "user_id": s["user_id"],
                    "day": s["day"],
                    "session_number": s["session_number"],
                    "started_at": s["started_at"],
                    "expires_at": s["expires_at"],
                    "ended_at": s["ended_at"],
                }
                for s in rows
            ]
        if "UPDATE free_sessions" in sql:
            # Lazy end (_mark_ended) passes only the id → ended_at = expires_at.
            # Explicit end (end_active_session) passes [ended_at, id] in SQL-text
            # order (psycopg is positional).
            if len(params) > 1:
                ended_at, row_id = params[0], params[1]
            else:
                ended_at, row_id = None, params[0]
            row = next((s for s in state["free_sessions"] if s["id"] == row_id), None)
            if row and row["ended_at"] is None:
                row["ended_at"] = ended_at if ended_at is not None else row["expires_at"]
            return []

        # ---- CostGuardService ----------------------------------------------
        if "INSERT INTO daily_cost" in sql:
            user_id, day, micros, cap_micros = params[0], params[1], params[2], params[3]
            row = next((r for r in state["daily_costs"] if r["user_id"] == user_id and r["day"] == day), None)
            if row:
                if row["spent"] + row["reserved"] + micros > cap_micros:
                    return []  # WHERE false → no row
                row["reserved"] += micros
                row["version"] += 1
                return [{"user_id": user_id}]
            state["daily_costs"].append({"user_id": user_id, "day": day, "spent": 0, "reserved": micros, "version": 1})
            return [{"user_id": user_id}]
        if "UPDATE daily_cost" in sql:
            # Settle passes 4 params in SQL-text order [spentDelta, reservedDelta,
            # user_id, day]; the release path passes 3 [reservedDelta, user_id, day].
            if len(params) >= 4:
                spent_delta, reserved_delta, user_id, day = params[0], params[1], params[2], params[3]
            else:
                reserved_delta, user_id, day = params[0], params[1], params[2]
                spent_delta = 0
            reserved_delta = reserved_delta or 0
            row = next((r for r in state["daily_costs"] if r["user_id"] == user_id and r["day"] == day), None)
            if row:
                row["spent"] += spent_delta
                row["reserved"] = max(0, row["reserved"] - reserved_delta)
                row["version"] += 1
            return []
        if "FROM daily_cost" in sql:
            user_id, day = params[0], params[1]
            row = next((r for r in state["daily_costs"] if r["user_id"] == user_id and r["day"] == day), None)
            return [{"spent_usd_micros": row["spent"], "reserved_usd_micros": row["reserved"]}] if row else []
        if "INSERT INTO global_cost" in sql:
            period_type, period, micros, cap_micros = params[0], params[1], params[2], params[3]
            row = next((r for r in state["global_costs"] if r["period_type"] == period_type and r["period"] == period), None)
            if row:
                if row["spent"] + row["reserved"] + micros > cap_micros:
                    return []
                row["reserved"] += micros
                row["version"] += 1
                return [{"period_type": period_type}]
            state["global_costs"].append({"period_type": period_type, "period": period, "spent": 0, "reserved": micros, "version": 1})
            return [{"period_type": period_type}]
        if "UPDATE global_cost" in sql:
            # Same text-order contract as daily_cost: settle = [spentDelta,
            # reservedDelta, period_type, period]; release = [reservedDelta,
            # period_type, period].
            if len(params) >= 4:
                spent_delta, reserved_delta, period_type, period = params[0], params[1], params[2], params[3]
            else:
                reserved_delta, period_type, period = params[0], params[1], params[2]
                spent_delta = 0
            reserved_delta = reserved_delta or 0
            row = next((r for r in state["global_costs"] if r["period_type"] == period_type and r["period"] == period), None)
            if row:
                row["spent"] += spent_delta
                row["reserved"] = max(0, row["reserved"] - reserved_delta)
                row["version"] += 1
            return []
        if "FROM global_cost" in sql:
            period_type, period = params[0], params[1]
            row = next((r for r in state["global_costs"] if r["period_type"] == period_type and r["period"] == period), None)
            return [{"spent_usd_micros": row["spent"], "reserved_usd_micros": row["reserved"]}] if row else []

        if "INSERT INTO ai_usage" in sql:
            state["ai_usage"].append(
                {
                    "user_id": params[0],
                    "session_id": params[1],
                    "provider": params[2],
                    "model": params[3],
                    "input_tokens": params[4],
                    "cached_input_tokens": params[5],
                    "output_tokens": params[6],
                    "total_tokens": params[7],
                    "estimated_cost_usd_micros": params[8],
                    "day": params[9],
                }
            )
            return []

        raise MemUnhandledQuery(f"MemoryDb: unhandled query: {sql}")

    return {"db": db, "state": state}
